Notification Service: pending intent doesnt start activity

Process: com.example.sfulounge, PID: 5712
                                                                                                    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.sfulounge/com.example.sfulounge.ui.messages.MessagesActivity}: java.lang.IllegalStateException: chat room members list is null
                                                                                                    	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2778)
                                                                                                    	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
                                                                                                    	at android.app.ActivityThread.-wrap11(Unknown Source:0)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                                                    	at android.os.Looper.loop(Looper.java:164)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:6494)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
                                                                                                    Caused by: java.lang.IllegalStateException: chat room members list is null
                                                                                                    	at com.example.sfulounge.ui.messages.MessagesActivity.onCreate(MessagesActivity.kt:45)
                                                                                                    	at android.app.Activity.performCreate(Activity.java:7009)
                                                                                                    	at android.app.Activity.performCreate(Activity.java:7000)
                                                                                                    	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
                                                                                                    	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
                                                                                                    	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
                                                                                                    	at android.app.ActivityThread.-wrap11(Unknown Source:0) 
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106) 
                                                                                                    	at android.os.Looper.loop(Looper.java:164) 
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:6494) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)

--- a/app/src/main/java/com/example/sfulounge/NotificationService.kt
+++ b/app/src/main/java/com/example/sfulounge/NotificationService.kt
@@ -44,6 +44,8 @@ class NotificationService : Service() {
     companion object {
         private const val NOTIFY_ID = -1
         private const val CHANNEL_ID = "sfu_lounge_messages"
+
+        const val ACTION_LOGOUT = "com.example.sfulounge.ACTION_LOGOUT"
     }
 
     override fun onCreate() {
@@ -82,10 +84,21 @@ class NotificationService : Service() {
         }
     }
 
+    private fun cancelAllNotifications() {
+        notificationManager.cancelAll()
+    }
+
     override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
         Log.d("Notification Service", "debug: onStartCommand called")
-        removeListener()
-        startListeningForMessages()
+        if (intent?.action == ACTION_LOGOUT) {
+            Log.d("Notification Service", "debug: action logout received. gracefully shutting down")
+            removeListener()
+            cancelAllNotifications()
+            stopSelf()
+        } else {
+            removeListener()
+            startListeningForMessages()
+        }
         return START_STICKY
     }
 
@@ -155,11 +168,15 @@ class NotificationService : Service() {
         val intent = Intent(this, MessagesActivity::class.java).apply {
             flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
             putExtra(MessagesActivity.INTENT_CHATROOM_ID, chatRoom.roomId)
-            putExtra(MessagesActivity.INTENT_MEMBER_IDS, chatRoom.members.toTypedArray())
+
+            val members = ArrayList<String>()
+            members.addAll(chatRoom.members)
+
+            putExtra(MessagesActivity.INTENT_MEMBER_IDS, members)
         }
         val pendingIntent = PendingIntent.getActivity(
             this, 0, intent,
-            PendingIntent.FLAG_IMMUTABLE
+            PendingIntent.FLAG_UPDATE_CURRENT
         )
 
         val notificationBuilder = NotificationCompat.Builder(this, CHANNEL_ID).apply {

--- a/app/src/main/java/com/example/sfulounge/ui/profile/ProfileFragment.kt
+++ b/app/src/main/java/com/example/sfulounge/ui/profile/ProfileFragment.kt
@@ -10,6 +10,7 @@ import android.widget.Toast
 import androidx.annotation.StringRes
 import androidx.fragment.app.Fragment
 import androidx.lifecycle.ViewModelProvider
+import com.example.sfulounge.NotificationService
 import com.example.sfulounge.R
 import com.example.sfulounge.data.model.Gender
 import com.example.sfulounge.data.model.User
@@ -170,7 +171,13 @@ class ProfileFragment : Fragment() {
     }
 
     private fun onLogoutClicked() {
+        // the service will stop gracefully
+        val serviceIntent = Intent(requireActivity(), NotificationService::class.java)
+        serviceIntent.action = NotificationService.ACTION_LOGOUT
+        requireActivity().startService(serviceIntent)
+
         profileViewModel.logout()
+
         Toast.makeText(
             requireActivity(),
             getString(R.string.success_message_logout),

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -46,7 +46,9 @@
     <string name="success_message_logout">Sign out successful!</string>
 
     <!--  confirm dialog  -->
-    <string name="confirm_logout">Are you sure you want to logout?</string>
+    <string name="confirm_logout">
+        Are you sure you want to logout? All notifications for this account will be cleared.
+    </string>
 
     <!--  dialog for messages  -->
     <string name="messages_upload_dialog_title">Upload Additional Items</string>

